File: examples/space-airlock/java/src/main/java/airlock/SpaceAirlock.java
package airlock;

/**
 * State machine that the shared gherclj features test against. Mirrors the
 * Ruby and Go implementations under examples/space-airlock/. Pure
 * production code — no test-framework dependency.
 */
public class SpaceAirlock {

    private static class Door {
        String position = "closed";
        String lock = "locked";
    }

    private static class Occupant {
        String name;
        String kind;
        boolean suit;
        boolean badgeValid;

        Occupant(String name, String kind) {
            this.name = name;
            this.kind = kind;
        }
    }

    private Occupant occupant;
    private String pressure = "pressurized";
    private final Door innerDoor = new Door();
    private final Door outerDoor = new Door();
    private boolean override;
    private String message = "";
    private String requestStatus = "";
    private String status = "";

    private void ensureOccupant() {
        if (occupant == null) {
            occupant = new Occupant(null, "crew");
        }
    }

    private void setOccupant(String name, String kind) {
        boolean suit = occupant != null && occupant.suit;
        boolean badge = occupant != null && occupant.badgeValid;
        Occupant next = new Occupant(name, kind);
        next.suit = suit;
        next.badgeValid = badge;
        occupant = next;
    }

    private Door door(String d) {
        return "inner".equals(d) ? innerDoor : outerDoor;
    }

    private void deny(String msg) {
        requestStatus = "denied";
        message = msg;
        status = "";
    }

    public void crewMemberInside(String name) { setOccupant(name, "crew"); }
    public void visitorInside()                { setOccupant("visitor", "visitor"); }

    public void wearingSuit(String name) {
        setOccupant(name, "crew");
        ensureOccupant();
        occupant.suit = true;
    }

    public void notWearingSuit(String name) {
        setOccupant(name, "crew");
        ensureOccupant();
        occupant.suit = false;
    }

    public void validBadge(String name) {
        setOccupant(name, "crew");
        ensureOccupant();
        occupant.badgeValid = true;
    }

    public void visitorInvalidBadge() {
        setOccupant("visitor", "visitor");
        ensureOccupant();
        occupant.badgeValid = false;
    }

    public void doorState(String d, String state) {
        Door door = door(d);
        switch (state) {
            case "open":
            case "closed":
                door.position = state;
                break;
            case "locked":
            case "unlocked":
                door.lock = state;
                break;
            default:
                throw new IllegalArgumentException("unknown door state: " + state);
        }
    }

    public void chamberPressure(String p)         { pressure = p; }
    public void emergencyOverrideEngaged()        { override = true; }

    public void requestExit(String name) {
        if (occupant == null || !occupant.badgeValid) {
            deny("Authorization required");
            return;
        }
        if (!occupant.suit) {
            deny("Suit required");
            return;
        }
        requestStatus = "approved";
        message = "";
        pressure = "depressurized";
        outerDoor.lock = "unlocked";
        innerDoor.lock = "locked";
        status = "cycle complete";
    }

    public void openDoor(String d) {
        Door door = door(d);
        if (override) {
            if ("inner".equals(d)) {
                door.lock = "unlocked";
                message = "Emergency override active";
            } else {
                door.lock = "locked";
                message = "Outer door lock preserved";
            }
            return;
        }
        if ("inner".equals(d)) {
            if ("depressurized".equals(pressure)) {
                door.lock = "locked";
                message = "Repressurize first";
            } else {
                door.lock = "unlocked";
                message = "";
            }
        } else {
            if ("pressurized".equals(pressure)) {
                door.lock = "locked";
                message = "Depressurize first";
            } else {
                door.lock = "unlocked";
                message = "";
            }
        }
    }

    public void depressurizationCommanded() {
        if ("open".equals(innerDoor.position)) {
            pressure = "pressurized";
            message = "Close inner door";
        } else {
            pressure = "depressurized";
            message = "";
        }
    }

    public void repressurizationCommanded() {
        if ("open".equals(outerDoor.position)) {
            pressure = "depressurized";
            message = "Close outer door";
        } else {
            pressure = "pressurized";
            message = "";
        }
    }

    public String getPressure()             { return pressure; }
    public String getDoorLock(String d)     { return door(d).lock; }
    public String getRequestStatus()        { return requestStatus; }
    public String getMessage()              { return message; }

    public String getStatus() {
        if (!status.isEmpty()) return status;
        if ("pressurized".equals(pressure)
                && "unlocked".equals(innerDoor.lock)
                && "locked".equals(outerDoor.lock)) {
            return "ready for boarding";
        }
        if ("depressurized".equals(pressure)
                && "locked".equals(innerDoor.lock)
                && "unlocked".equals(outerDoor.lock)) {
            return "ready for exit";
        }
        return "idle";
    }
}

File: examples/space-airlock/csharp/src/SpaceAirlock/SpaceAirlock.cs
namespace SpaceAirlock
{
public enum OccupantKind
{
    Crew,
    Visitor,
}

public enum Pressure
{
    Pressurized,
    Depressurized,
}

public enum Door
{
    Inner,
    Outer,
}

public enum DoorPosition
{
    Open,
    Closed,
}

public enum DoorLock
{
    Locked,
    Unlocked,
}

public abstract record DoorStateValue
{
    private DoorStateValue() { }
    public sealed record Position(DoorPosition Value) : DoorStateValue;
    public sealed record Lock(DoorLock Value) : DoorStateValue;
}

internal sealed record Occupant(string Name, OccupantKind Kind, bool Suit, bool BadgeValid);
internal sealed record DoorState(DoorPosition Position, DoorLock Lock);
internal enum RequestStatus { Approved, Denied }

public sealed class SpaceAirlock
{
    private Occupant? _occupant;
    private Pressure _pressure = Pressure.Pressurized;
    private DoorState _innerDoor = new(DoorPosition.Closed, DoorLock.Locked);
    private DoorState _outerDoor = new(DoorPosition.Closed, DoorLock.Locked);
    private bool _emergencyOverride;
    private string? _message;
    private RequestStatus? _requestStatus;
    private string? _status;

    public void CrewMemberInside(string name) => SetOccupant(name, OccupantKind.Crew);
    public void VisitorInside() => SetOccupant("visitor", OccupantKind.Visitor);

    public void WearingSuit(string name)
    {
        SetOccupant(name, OccupantKind.Crew);
        SetSuit(true);
    }

    public void NotWearingSuit(string name)
    {
        SetOccupant(name, OccupantKind.Crew);
        SetSuit(false);
    }

    public void ValidBadge(string name)
    {
        SetOccupant(name, OccupantKind.Crew);
        SetBadge(true);
    }

    public void VisitorInvalidBadge()
    {
        SetOccupant("visitor", OccupantKind.Visitor);
        SetBadge(false);
    }

    public void SetDoorState(Door door, DoorStateValue state)
    {
        var current = GetDoorState(door);
        var updated = state switch
        {
            DoorStateValue.Position p => current with { Position = p.Value },
            DoorStateValue.Lock l => current with { Lock = l.Value },
            _ => current,
        };
        SetDoorState(door, updated);
    }

    public void SetPressure(Pressure pressure) => _pressure = pressure;

    public void EngageEmergencyOverride() => _emergencyOverride = true;

    public void RequestExit(string _)
    {
        var occupant = _occupant ?? DefaultOccupant();
        if (!occupant.BadgeValid)
        {
            Deny("Authorization required");
            return;
        }
        if (!occupant.Suit)
        {
            Deny("Suit required");
            return;
        }

        _requestStatus = RequestStatus.Approved;
        _message = null;
        _pressure = Pressure.Depressurized;
        _outerDoor = _outerDoor with { Lock = DoorLock.Unlocked };
        _innerDoor = _innerDoor with { Lock = DoorLock.Locked };
        _status = "cycle complete";
    }

    public void CommandOpenDoor(Door door)
    {
        if (_emergencyOverride)
        {
            if (door == Door.Inner)
            {
                _innerDoor = _innerDoor with { Lock = DoorLock.Unlocked };
                _message = "Emergency override active";
            }
            else
            {
                _outerDoor = _outerDoor with { Lock = DoorLock.Locked };
                _message = "Outer door lock preserved";
            }
            return;
        }

        switch (door)
        {
            case Door.Inner when _pressure == Pressure.Depressurized:
                _innerDoor = _innerDoor with { Lock = DoorLock.Locked };
                _message = "Repressurize first";
                break;
            case Door.Inner:
                _innerDoor = _innerDoor with { Lock = DoorLock.Unlocked };
                _message = null;
                break;
            case Door.Outer when _pressure == Pressure.Pressurized:
                _outerDoor = _outerDoor with { Lock = DoorLock.Locked };
                _message = "Depressurize first";
                break;
            case Door.Outer:
                _outerDoor = _outerDoor with { Lock = DoorLock.Unlocked };
                _message = null;
                break;
}

}

    public void Depressurize()
    {
        if (_innerDoor.Position == DoorPosition.Open)
        {
            _pressure = Pressure.Pressurized;
            _message = "Close inner door";
        }
        else
        {
            _pressure = Pressure.Depressurized;
            _message = null;
        }
    }

    public void Repressurize()
    {
        if (_outerDoor.Position == DoorPosition.Open)
        {
            _pressure = Pressure.Depressurized;
            _message = "Close outer door";
        }
        else
        {
            _pressure = Pressure.Pressurized;
            _message = null;
        }
    }

    public Pressure Pressure => _pressure;
    public DoorLock DoorLockFor(Door door) => GetDoorState(door).Lock;
    public string? Message => _message;
    public bool RequestDenied => _requestStatus == RequestStatus.Denied;

    public string Status => _status
        ?? (_pressure, _innerDoor.Lock, _outerDoor.Lock) switch
        {
            (Pressure.Pressurized, DoorLock.Unlocked, DoorLock.Locked) => "ready for boarding",
            (Pressure.Depressurized, DoorLock.Locked, DoorLock.Unlocked) => "ready for exit",
            _ => "idle",
        };

    private void SetOccupant(string name, OccupantKind kind)
    {
        var current = _occupant ?? DefaultOccupant();
        _occupant = new Occupant(name, kind, current.Suit, current.BadgeValid);
    }

    private void SetSuit(bool suit)
    {
        var current = _occupant ?? DefaultOccupant();
        _occupant = current with { Suit = suit };
    }

    private void SetBadge(bool badgeValid)
    {
        var current = _occupant ?? DefaultOccupant();
        _occupant = current with { BadgeValid = badgeValid };
    }

    private void Deny(string message)
    {
        _requestStatus = RequestStatus.Denied;
        _message = message;
        _status = null;
    }

    private DoorState GetDoorState(Door door) => door == Door.Inner ? _innerDoor : _outerDoor;

    private void SetDoorState(Door door, DoorState state)
    {
        if (door == Door.Inner)
            _innerDoor = state;
        else
            _outerDoor = state;
    }

private static Occupant DefaultOccupant() => new(string.Empty, OccupantKind.Crew, false, false);
}

}
